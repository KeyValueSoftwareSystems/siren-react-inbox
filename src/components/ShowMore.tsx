import React from "react";

import "../styles/showMore.css";
import type { LoadMoreProps } from "../types";

const ShowMoreButton = (props: LoadMoreProps) => {
  const {styles, customComponent, onClick, loadMoreLabel } = props;

  if(customComponent) 
    return (
      <button
        className="siren-sdk-panel-load-more-custom"
        onClick={onClick}
        data-testid="load-more"
      >
        {customComponent}
      </button>
    )

  return (
    <button
      className="siren-sdk-panel-load-more"
      onClick={onClick}
      data-testid="load-more"
      style={styles.loadMoreButton}
    >
      {loadMoreLabel || 'Load More'}
    </button>
  );
};

export default ShowMoreButton;
