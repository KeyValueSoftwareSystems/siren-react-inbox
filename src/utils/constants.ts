export const COLORS = {
  light: {
    primaryColor: "#F56630",
    highlightedCardColor: "#FFECE5",
    textColor: "#344054",
    neutralColor: "#FFFFFF",
    borderColor: "#D0D5DD",
    dateColor: "#34405499",
    deleteIcon: "#34405499",
    timerIcon: "#667185",
    clearAllIcon: "#667185",
    windowShadowColor: "#D0D5DD",
    iconColor: "#F7F9FC",
    notificationIcon: "#232326",
  },
  dark: {
    primaryColor: "#F56630",
    highlightedCardColor: "#2E2D30",
    textColor: "#FFFFFF",
    neutralColor: "#232326",
    borderColor: "#344054",
    dateColor: "#98A2B3",
    deleteIcon: "#D0D5DD",
    timerIcon: "#98A2B3",
    clearAllIcon: "#D0D5DD",
    windowShadowColor: "#232326",
    iconColor: "#38383D",
    notificationIcon: "#FFFFFF",
  },
};

export enum ThemeMode {
  DARK = "dark",
  LIGHT = "light",
}

export enum LogLevel {
  INFO,
  ERROR,
}

export enum BadgeType {
  NONE = "none",
  DOT = "dot",
  DEFAULT = "default",
}

export const levelLogFns = {
  [LogLevel.INFO]: console.log,
  [LogLevel.ERROR]: console.error,
};

export const defaultBadgeStyle = {
  size: 15,
  color: "#FF0000",
  textColor: "#FFFFFF",
  textSize: 10,
  linHeight: "14px",
  top: "inherit",
  left: "inherit",
};

export enum eventTypes {
  MARK_ITEM_AS_VIEWED = "MARK_ITEM_AS_VIEWED",
  MARK_ITEM_AS_READ = "MARK_ITEM_AS_READ",
  MARK_ALL_AS_READ = "MARK_ALL_AS_READ",
  DELETE_ITEM = "DELETE_ITEM",
  DELETE_ALL_ITEM = "DELETE_ALL_ITEM",
  NEW_NOTIFICATIONS = "NEW_NOTIFICATIONS",
  UPDATE_NOTIFICATIONS_COUNT = "UPDATE_NOTIFICATIONS_COUNT",
  RESET_NOTIFICATIONS = "RESET_NOTIFICATIONS",
  RESET_NOTIFICATIONS_COUNT = "RESET_NOTIFICATIONS_COUNT",
}

export enum events {
  NOTIFICATION_LIST_EVENT = "NOTIFICATION_LIST_EVENT",
  NOTIFICATION_COUNT_EVENT = "NOTIFICATION_COUNT_EVENT",
}

export enum updateNotificationsTypes {
  MARK_ITEM_AS_VIEWED = "MARK_ITEM_AS_VIEWED",
  MARK_ITEM_AS_READ = "MARK_ITEM_AS_READ",
  MARK_ALL_AS_READ = "MARK_ALL_AS_READ",
  DELETE_ITEM = "DELETE_ITEM",
  DELETE_ALL_ITEM = "DELETE_ALL_ITEM",
}

export enum sirenReducerTypes {
  SET_NOTIFICATIONS = "SET_NOTIFICATIONS",
  NEW_NOTIFICATIONS = "NEW_NOTIFICATIONS",
  SET_UN_VIEWED_NOTIFICATION_COUNT = "SET_UN_VIEWED_NOTIFICATION_COUNT",
  SET_SIREN_CORE = "SET_SIREN_CORE",
}

export const LIST_EMPTY_TEXT = "No new notifications";
export const LIST_EMPTY_SUB_TEXT = "Check back later for updates and alerts.";
export const ERROR_TEXT = "Oops! Something went wrong";
export const ERROR_SUB_TEXT =
  "Could not load the notifications. Please refresh the page.";
export const DEFAULT_WINDOW_TITLE = "Notifications";
export const RETRY_BUTTON_LABEL = "Retry";
export const CLEAR_ALL_LABEL = "Clear All";
export const IN_APP_RECIPIENT_UNAUTHENTICATED = 'IN_APP_RECIPIENT_UNAUTHENTICATED';
export const TOKEN_VERIFICATION_PENDING ='TOKEN_VERIFICATION_PENDING';
export const MAXIMUM_RETRY_COUNT = 3;

export const errorMap = {
  SIREN_OBJECT_NOT_FOUND: {
    Type: "ERROR",
    Code: "SIREN_OBJECT_NOT_FOUND",
    Message: "Siren Object Not found",
  },
  MISSING_PARAMETER: {
    Type: "ERROR",
    Code: "MISSING_PARAMETER",
    Message: "Missing Parameter",
  },
};

export enum VerificationStatus {
  PENDING = "PENDING",
  SUCCESS = "SUCCESS",
  FAILED = "FAILED"
}