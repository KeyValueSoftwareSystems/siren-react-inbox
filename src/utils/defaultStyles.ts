
const defaultStyles = {
  notificationIcon: {
    size: 30,
  },
  window: {
    width: 500,
    borderRadius: "10px",
    borderWidth: "0px",
    borderStyle: "solid"
  },
  windowHeader: {
    height: 64,
    titleFontWeight: "600",
    titleSize: 20,
    closeIconSize: 20,
    titlePadding: 0,
  },
  windowContainer: {
    padding: 0,
    height: 700,
  },
  notificationCard: {
    padding: 10,
    borderWidth: 0.5,
    avatarSize: 40,
    titleFontWeight: "600",
    titleSize: 16,
    descriptionSize: 14,
    dateSize: 12,
  },
  loadMoreButton: {
    fontSize: 14,
    fontWeight: '600'
  },
  badgeStyle: {
    size: 15,
    textSize: 10,
  },
  deleteIcon:{
    size: 16
  },
  timerIcon:{
    size: 12
  },
  clearAllIcon:{
    size: 18
  }
}

export default defaultStyles;
