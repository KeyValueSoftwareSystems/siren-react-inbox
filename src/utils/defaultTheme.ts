import { COLORS, ThemeMode } from "./constants";

const defaultTheme = {
  light: {
    colors: {
      primaryColor: COLORS[ThemeMode.LIGHT].primaryColor,
      textColor: COLORS[ThemeMode.LIGHT].textColor,
      neutralColor: COLORS[ThemeMode.LIGHT].neutralColor,
      borderColor: COLORS[ThemeMode.LIGHT].borderColor,
      highlightedCardColor: COLORS[ThemeMode.LIGHT].highlightedCardColor,

      dateColor: COLORS[ThemeMode.LIGHT].dateColor,
      deleteIcon: COLORS[ThemeMode.LIGHT].deleteIcon,
      timerIcon: COLORS[ThemeMode.LIGHT].timerIcon,
      clearAllIcon: COLORS[ThemeMode.LIGHT].clearAllIcon,
    },
    window: {
      borderColor: COLORS[ThemeMode.LIGHT].borderColor,
      shadowColor: COLORS[ThemeMode.DARK].windowShadowColor,
    },
    windowHeader: {
      background: COLORS[ThemeMode.LIGHT].neutralColor,
      titleColor: COLORS[ThemeMode.LIGHT].textColor,
      headerActionColor: COLORS[ThemeMode.LIGHT].textColor,
      borderColor: COLORS[ThemeMode.LIGHT].borderColor,
    },
    windowContainer: {
      background: COLORS[ThemeMode.LIGHT].neutralColor,
    },
    notificationCard: {
      borderColor: COLORS[ThemeMode.LIGHT].borderColor,
      titleColor: COLORS[ThemeMode.LIGHT].textColor,
      background: "transparent",
      subTitleColor: COLORS[ThemeMode.LIGHT].textColor,
      descriptionColor: COLORS[ThemeMode.LIGHT].textColor,
    },
    loadMoreButton: {
      color: COLORS[ThemeMode.LIGHT].primaryColor,
      background: COLORS[ThemeMode.LIGHT].neutralColor,
    },
    deleteIcon: {
      color: COLORS[ThemeMode.LIGHT].deleteIcon,
    },
    clearIcon: {
      color: COLORS[ThemeMode.LIGHT].clearAllIcon,
    },
    timerIcon : {
      color: COLORS[ThemeMode.LIGHT].timerIcon,
    },
    loader:{
      backgroundImage: 'linear-gradient(to right, #F0f0f0 0%, #E0E0E0 50%, #F0F0F0 100%)',
    }
  },
  dark: {
    colors: {
      primaryColor: COLORS[ThemeMode.DARK].primaryColor,
      textColor: COLORS[ThemeMode.DARK].textColor,
      neutralColor: COLORS[ThemeMode.DARK].neutralColor,
      borderColor: COLORS[ThemeMode.DARK].borderColor,
      highlightedCardColor: COLORS[ThemeMode.DARK].highlightedCardColor,
      dateColor: COLORS[ThemeMode.DARK].dateColor,
      deleteIcon: COLORS[ThemeMode.DARK].deleteIcon,
      timerIcon: COLORS[ThemeMode.DARK].timerIcon,
      clearAllIcon: COLORS[ThemeMode.DARK].clearAllIcon,
    },
    window: {
      borderColor: COLORS[ThemeMode.DARK].borderColor,
      shadowColor: COLORS[ThemeMode.DARK].windowShadowColor,
    },
    windowHeader: {
      background: COLORS[ThemeMode.DARK].neutralColor,
      titleColor: COLORS[ThemeMode.DARK].textColor,
      headerActionColor: COLORS[ThemeMode.DARK].textColor,
      borderColor: COLORS[ThemeMode.DARK].borderColor,
    },
    windowContainer: {
      background: COLORS[ThemeMode.DARK].neutralColor,
    },
    notificationCard: {
      borderColor: COLORS[ThemeMode.DARK].borderColor,
      background: "transparent",
      titleColor: COLORS[ThemeMode.DARK].textColor,
      subTitleColor: COLORS[ThemeMode.DARK].textColor,
      descriptionColor: COLORS[ThemeMode.DARK].textColor,
    },
    loadMoreButton: {
      color: COLORS[ThemeMode.LIGHT].primaryColor,
      background: COLORS[ThemeMode.DARK].neutralColor,
    },
    deleteIcon: {
      color: COLORS[ThemeMode.DARK].deleteIcon,
    },
    clearIcon: {
      color: COLORS[ThemeMode.DARK].clearAllIcon,
    },
    timerIcon : {
      color: COLORS[ThemeMode.DARK].timerIcon,
    },
    loader:{
      backgroundImage: 'linear-gradient(to right, #49494A 0%, #535354 50%, #49494A 100%)',
    }
  },
};

export default defaultTheme;
