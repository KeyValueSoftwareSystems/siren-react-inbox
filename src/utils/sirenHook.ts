import PubSub from "pubsub-js";

import { errorMap, events, eventTypes } from "./constants";
import { useSirenContext } from "../components/SirenProvider";

const useSiren = () => {
  const { siren, id: providerInstanceId } = useSirenContext();

  const markAsReadById = async (id: string) => {
    if (siren)
      if (id?.length > 0) {
        const response = await siren?.markAsReadById(id);

        if (response?.data) {
          const payload = { id, action: eventTypes.MARK_ITEM_AS_READ };

          PubSub.publish(
            `${events.NOTIFICATION_LIST_EVENT}${providerInstanceId}`,
            JSON.stringify(payload)
          );
        }

        return response;
      } else {
        return { error: errorMap.MISSING_PARAMETER };
      }

    return { error: errorMap.SIREN_OBJECT_NOT_FOUND };
  };

  const markAsReadByDate = async (untilDate: string, category?: string) => {
    if (siren && untilDate) {
      const params = { startDate: untilDate, category };
      const response = await siren?.markAsReadByDate(params);

      if (response?.data) {
        const payload = { action: eventTypes.MARK_ALL_AS_READ };

        PubSub.publish(`${events.NOTIFICATION_LIST_EVENT}${providerInstanceId}`, JSON.stringify(payload));
      }

      return response;
    }

    return { error: errorMap.SIREN_OBJECT_NOT_FOUND };
  };
  
  const deleteById = async (id: string, shouldUpdateList: boolean = true) => {
    if (siren)
      if (id?.length > 0) {
        const response = await siren?.deleteById(id);

        if (response?.data && shouldUpdateList) {
          const payload = { id, action: eventTypes.DELETE_ITEM };

          PubSub.publish(
            `${events.NOTIFICATION_LIST_EVENT}${providerInstanceId}`,
            JSON.stringify(payload)
          );
        }

        return response;
      } else {
        return { error: errorMap.MISSING_PARAMETER };
      }

    return { error: errorMap.SIREN_OBJECT_NOT_FOUND };
  };

  const deleteByDate = async (untilDate: string, category?: string) => {
    if (siren && untilDate) {
      const params = { startDate: untilDate, category };
      const response = await siren?.deleteByDate(params);

      if (response?.data) {
        const payload = { action: eventTypes.DELETE_ALL_ITEM };

        PubSub.publish(`${events.NOTIFICATION_LIST_EVENT}${providerInstanceId}`, JSON.stringify(payload));
      }

      return response;
    }

    return { error: errorMap.SIREN_OBJECT_NOT_FOUND };
  };

  const markAllAsViewed = async (untilDate: string) => {
    if (siren && untilDate) {
      const response = await siren?.markAllAsViewed(untilDate);

      if (response && response.data) {
        const payload = {
          notificationsCount: 0,
          action: eventTypes.UPDATE_NOTIFICATIONS_COUNT,
        };

        PubSub.publish(
          `${events.NOTIFICATION_COUNT_EVENT}${providerInstanceId}`,
          JSON.stringify(payload)
        );
      }

      return response;
    }

    return { error: errorMap.SIREN_OBJECT_NOT_FOUND };
  };

  return {
    markAsReadByDate,
    markAsReadById,
    deleteById,
    deleteByDate,
    markAllAsViewed,
  };
};

export default useSiren;
