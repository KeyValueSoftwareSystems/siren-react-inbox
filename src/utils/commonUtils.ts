import type { RefObject } from "react";

import type {
  ActionResponse,
  NotificationDataType,
  NotificationsApiResponse,
} from "@sirenapp/js-sdk/dist/esm/types";

import {
  defaultBadgeStyle,
  eventTypes,
  LogLevel,
  ThemeMode,
} from "./constants";
import { default as DefaultStyle } from "./defaultStyles";
import { default as DefaultTheme } from "./defaultTheme";
import type {
  CustomStyle,
  DimensionValue,
  SirenStyleProps,
  ThemeProps,
} from "../types";

type FetchParams = {
  size: number;
  start?: string;
  end?: string;
  sort?: "createdAt" | "updatedAt";
};

export const generateElapsedTimeText = (timeString: string) => {
  const currentTime = new Date().getTime();
  const targetTime = new Date(timeString).getTime();
  const millisecondsDiff = currentTime - targetTime;

  const seconds = Math.floor(millisecondsDiff / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);
  const months = Math.floor(days / 30);
  const years = Math.floor(days / 365);

  if (millisecondsDiff < 60000) return "Just now";
  else if (minutes < 60)
    return minutes === 1 ? "1 minute ago" : `${minutes} minutes ago`;
  else if (hours < 24) return hours === 1 ? "1 hour ago" : `${hours} hours ago`;
  else if (days < 30) return days === 1 ? "1 day ago" : `${days} days ago`;
  else if (months < 12)
    return months === 1 ? "1 month ago" : `${months} months ago`;
  else return years === 1 ? "1 year ago" : `${years} years ago`;
};

export const updateNotifications = (
  eventData: {
    id?: string;
    action: string;
    newNotifications?: NotificationDataType[];
    unreadCount?: number;
  },
  notifications: NotificationDataType[]
): NotificationDataType[] => {
  let updatedNotifications: NotificationDataType[] = [];

  switch (eventData.action) {
    case eventTypes.MARK_ITEM_AS_READ:
      updatedNotifications = notifications.map((item) =>
        item.id === eventData.id ? { ...item, isRead: true } : item
      );
      break;
    case eventTypes.MARK_ALL_AS_READ:
      updatedNotifications = notifications.map((item) => ({
        ...item,
        isRead: true,
      }));
      break;
    case eventTypes.DELETE_ITEM:
      updatedNotifications = notifications.filter(
        (item) => item.id !== eventData.id
      );
      break;
    case eventTypes.DELETE_ALL_ITEM:
      updatedNotifications = [];
      break;
    case eventTypes.NEW_NOTIFICATIONS: {
      const newNotifications: NotificationDataType[] =
        eventData?.newNotifications || [];

      updatedNotifications = [...newNotifications, ...notifications];
      break;
    }
    case eventTypes.RESET_NOTIFICATIONS: {
      updatedNotifications = [];
      break;
    }
    default:
      break;
  }

  return updatedNotifications;
};

export const renderAsImage = (icon: JSX.Element | string): boolean =>
  typeof icon === "undefined" || typeof icon === "string";

export const logger = {
  log: async (level: LogLevel.INFO | LogLevel.ERROR, message: string) => {
    const timestamp = new Date().toISOString();
    const levelString = LogLevel[level].toUpperCase();

    console.log(`[${timestamp}] [${levelString}] ${message}`);
  },
  error: function (error: string) {
    this.log(LogLevel.ERROR, error);
  },
  info: function (message: string) {
    this.log(LogLevel.INFO, message);
  },
};

export const applyTheme = (
  theme: ThemeProps = {},
  mode: ThemeMode = ThemeMode.DARK,
  customStyle: CustomStyle = {}
): SirenStyleProps => {
  const windowBorderRadius =
    customStyle.window?.borderRadius || DefaultStyle.window.borderRadius;

  return {
    container: {
      maxWidth: customStyle.window?.width || "100",
    },
    windowShadow: {
      boxShadow: `${
        theme.colors?.windowShadowColor || DefaultTheme[mode].window.shadowColor
      } 0px 8px 24px`,
    },
    windowTopBorder: {
      borderTopLeftRadius: windowBorderRadius,
      borderTopRightRadius: windowBorderRadius,
    },
    windowBottomBorder: {
      borderBottomLeftRadius: windowBorderRadius,
      borderBottomRightRadius: windowBorderRadius,
    },
    contentContainer: {
      backgroundColor:
        theme.windowContainer?.background ||
        theme.colors?.neutralColor ||
        DefaultTheme[mode].windowContainer.background,
      padding:
        customStyle.windowContainer?.padding ||
        DefaultStyle.windowContainer.padding,
      borderRadius: `0 0 ${
        customStyle.window?.borderRadius || DefaultStyle.window.borderRadius
      } ${
        customStyle.window?.borderRadius || DefaultStyle.window.borderRadius
      }`,
    },
    body: {
      overflow: "auto",
      height: customStyle.windowContainer?.contentHeight || "700px",
    },
    headerContainer: {
      alignItems: "center",
      display: "flex",
      margin: 0,
      lineHeight: "28px",
      backgroundColor:
        theme.windowHeader?.background ||
        theme.colors?.neutralColor ||
        DefaultTheme[mode].windowHeader.background,
      borderBottom: `${customStyle.windowHeader?.borderWidth || DefaultStyle.windowHeader.borderWidth} solid`,
      borderColor: theme.colors?.borderColor || DefaultTheme[mode].colors.borderColor,
      height:
        customStyle.windowHeader?.height || DefaultStyle.windowHeader.height,
    },
    headerTitle: {
      color:
        theme.windowHeader?.titleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].windowHeader.titleColor,
      fontSize:
        customStyle.windowHeader?.titleSize ||
        DefaultStyle.windowHeader.titleSize,
      fontWeight:
        customStyle.windowHeader?.titleFontWeight ||
        DefaultStyle.windowHeader.titleFontWeight,
      paddingLeft:
        customStyle.windowHeader?.titlePadding ||
        DefaultStyle.windowHeader.titlePadding,
    },
    headerAction: {
      color:
        theme.windowHeader?.headerActionColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].windowHeader.headerActionColor,
    },
    defaultCardContainer: {
      backgroundColor:
        theme.notificationCard?.background ||
        DefaultTheme[mode].notificationCard.background,
      padding:
        customStyle.notificationCard?.padding ||
        DefaultStyle.notificationCard.padding,
      borderBottom: `${
        customStyle.notificationCard?.borderWidth ||
        DefaultStyle.notificationCard.borderWidth
      }px solid`,
      borderColor:
        theme.notificationCard?.borderColor ||
        theme.colors?.borderColor ||
        DefaultTheme[mode].notificationCard.borderColor,
    },
    cardIconRound: {
      width:
        customStyle.notificationCard?.avatarSize ||
        DefaultStyle.notificationCard.avatarSize,
      height:
        customStyle.notificationCard?.avatarSize ||
        DefaultStyle.notificationCard.avatarSize,
      borderRadius:
        (customStyle.notificationCard?.avatarSize ||
          DefaultStyle.notificationCard.avatarSize) / 2,
      overflow: "hidden",
      backgroundColor:
        theme.colors?.borderColor || DefaultTheme[mode].colors.borderColor,
    },
    cardTitle: {
      color:
        theme.notificationCard?.titleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.titleColor,
      fontSize:
        customStyle.notificationCard?.titleSize ||
        DefaultStyle.notificationCard.titleSize,
      fontWeight:
        customStyle.notificationCard?.titleFontWeight ||
        DefaultStyle.notificationCard.titleFontWeight,
    },
    cardSubTitle: {
      color:
        theme.notificationCard?.subTitleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.subTitleColor,
      fontSize:
        customStyle.notificationCard?.subTitleSize || DefaultStyle.notificationCard.subTitleSize,
      fontWeight:
        customStyle.notificationCard?.subTitleFontWeight ||
        DefaultStyle.notificationCard.subTitleFontWeight,
    },
    activeCardMarker: {
      backgroundColor:
        theme.colors?.highlightedCardColor ||
        DefaultTheme[mode].colors?.highlightedCardColor,
      border:
        theme.colors?.primaryColor || DefaultTheme[mode].colors?.primaryColor,
    },
    cardDescription: {
      color:
        theme.notificationCard?.descriptionColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.descriptionColor,
      fontSize:
        customStyle.notificationCard?.descriptionSize ||
        DefaultStyle.notificationCard.descriptionSize,
      fontWeight:
      customStyle.notificationCard?.descriptionFontWeight ||
      DefaultStyle.notificationCard.descriptionFontWeight,
    },
    dateStyle: {
      color: theme.colors?.dateColor || DefaultTheme[mode].colors.dateColor,
      fontSize:
        customStyle.notificationCard?.dateSize ||
        DefaultStyle.notificationCard.dateSize,
      lineHeight: "16px",
    },
    emptyText: {
      color: theme.colors?.textColor || DefaultTheme[mode].colors.textColor,
    },
    errorText: {
      color: theme.colors?.textColor || DefaultTheme[mode].colors.textColor,
    },
    clearIcon: {
      color: theme.colors?.clearAllIcon || DefaultTheme[mode].clearIcon.color,
      size: customStyle.clearAllIcon?.size || DefaultStyle.clearAllIcon.size,
    },
    timerIcon: {
      color: theme.colors?.timerIcon || DefaultTheme[mode].timerIcon.color,
      size: customStyle.timerIcon?.size || DefaultStyle.timerIcon.size,
    },
    notificationIcon: {
      size:
        customStyle?.notificationIcon?.size ||
        DefaultStyle?.notificationIcon?.size,
    },
    deleteIcon: {
      color: theme.colors?.deleteIcon || DefaultTheme[mode].colors.deleteIcon,
      size: customStyle.deleteIcon?.size || DefaultStyle.deleteIcon.size,
    },
    loadMoreButton: {
      color:
        theme?.loadMoreButton?.color ||
        theme.colors?.primaryColor ||
        DefaultTheme[mode]?.loadMoreButton?.color,
      backgroundColor:
        theme?.loadMoreButton?.background ||
        DefaultTheme[mode]?.loadMoreButton?.background,
      fontSize:
        customStyle?.loadMoreButton?.fontSize ||
        DefaultStyle?.loadMoreButton?.fontSize,
      fontWeight:
        customStyle.loadMoreButton?.fontWeight ||
        DefaultStyle.loadMoreButton.fontWeight,
    },
    loader: {
      backgroundImage: DefaultTheme[mode].loader.backgroundImage,
      borderColor:
        theme.colors?.primaryColor || DefaultTheme[mode].colors.primaryColor,
    },
    badgeStyle: {
      borderRadius: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      minWidth: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      height: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      backgroundColor: theme.badgeStyle?.color || defaultBadgeStyle.color,
      top:  customStyle?.badgeStyle?.top ? `${customStyle.badgeStyle.top}px` : defaultBadgeStyle.top,
      right:  customStyle?.badgeStyle?.right ? `${customStyle.badgeStyle.right}px` : defaultBadgeStyle.right,    
    },
    badgeTextStyle: {
      color: theme.badgeStyle?.textColor || defaultBadgeStyle.textColor,
      fontSize: customStyle.badgeStyle?.textSize || defaultBadgeStyle.textSize,
    },
    infiniteLoader: {
      border: `3px solid ${
        theme.colors?.infiniteLoader ||
        theme.colors?.primaryColor ||
        DefaultTheme[mode].colors.primaryColor
      }`,
    },
  };
};

export const isEmptyArray = (array: Array<NotificationDataType> = []) => {
  return array && array?.length === 0;
};
export const mergeArrays = (
  array1: Array<NotificationDataType> = [],
  array2: Array<NotificationDataType> = []
) => {
  if (array1 && array2) return [...array1, ...array2];
  if (array1) return array1;

  return array2;
};

export const isValidResponse = (
  response: NotificationsApiResponse | ActionResponse
): boolean => !!response?.data;

export const filterDataProperty = (
  response: NotificationsApiResponse
): NotificationDataType[] | null => {
  if (!response.data) return null;

  return response.data;
};

export const generateFilterParams = (
  data: NotificationDataType[],
  fromStart: boolean,
  itemsPerPage: number
): FetchParams => {
  let params: FetchParams = { size: itemsPerPage, sort: "createdAt" };

  if (data.length > 0)
    if (fromStart) params = { ...params, start: data[0].createdAt };
    else params = { ...params, end: data[data.length - 1].createdAt };

  return params;
};

export const calculateModalPosition = (
  iconRef: RefObject<HTMLDivElement>,
  window: Window,
  containerWidth: DimensionValue
) => {
  if (iconRef.current) {
    const iconRect = iconRef.current.getBoundingClientRect();
    const screenWidth = window.innerWidth;
    const spaceRight = screenWidth - iconRect.x;
    const spaceLeft = iconRect.x;
    let modalWidth = calculateModalWidth(containerWidth);

    const centerPosition =
      Math.min(spaceLeft, spaceRight) + Math.abs(spaceLeft - spaceRight) / 2;

    if (window.innerWidth <= modalWidth) modalWidth = window.innerWidth - 40;

    if (spaceRight > modalWidth) {
      return {
        left: `0px`,
      };
    } else if (spaceLeft > modalWidth) {
      const rightPosition = spaceRight < modalWidth ? "30px" : null;

      return {
        ...(rightPosition && { right: rightPosition }),
      };
    } else if (
      spaceLeft < modalWidth &&
      spaceRight < modalWidth &&
      spaceLeft > spaceRight
    ) {
      return { right: "30px" };
    } else {
      return { left: `-${centerPosition - 40}px` };
    }
  }
};

export const calculateModalWidth = (containerWidth: DimensionValue): number => {
  let modalWidth = 500;

  if (typeof containerWidth === "string")
    modalWidth = parseInt(containerWidth.slice(0, -2)) + 40;
  else if (typeof containerWidth === "number") modalWidth = containerWidth + 40;

  return modalWidth;
};

export const hexToRgba = (hex: string, alpha: number) => {
  const [r, g, b] = hex.match(/\w\w/g)?.map((x) => parseInt(x, 16)) ?? [];

  return `rgba(${r},${g},${b},${alpha})`;
};

export const debounce = <F extends (...args: unknown[]) => void>(
  func: F,
  delay: number
) => {
  let timerId: ReturnType<typeof setTimeout>;

  return (...args: Parameters<F>): void => {
    clearTimeout(timerId);
    timerId = setTimeout(() => {
      func(...args);
    }, delay);
  };
};
