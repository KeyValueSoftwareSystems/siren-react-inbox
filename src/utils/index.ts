import * as CommonUtils from './commonUtils';
import * as Constants from './constants';
import DefaultStyle from './defaultStyles'
import DefaultTheme from './defaultTheme';
import useSiren from './sirenHook';

export {
  Constants,
  DefaultTheme,
  useSiren,
  CommonUtils,
  DefaultStyle
};