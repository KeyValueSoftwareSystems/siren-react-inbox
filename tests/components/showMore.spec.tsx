import React from "react";

import {  queryByAttribute, render } from "@testing-library/react";

import ShowMore from "../../src/components/ShowMore";
import { applyTheme } from "../../src/utils/commonUtils";

const mockfn = jest.fn();

// Mock the CSS file to avoid Jest error
jest.mock('../../src/styles/showMore.css', () => ({}));

const style = applyTheme();

test("matches snapshot", () => {
  const { asFragment } = render(
    <ShowMore
      loadMoreLabel={"dummy-title"}
      styles={style}
      customComponent={<div></div>}
      onClick={mockfn}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});