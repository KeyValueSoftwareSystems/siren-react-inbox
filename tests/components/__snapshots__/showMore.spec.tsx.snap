// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="siren-sdk-panel-load-more-custom"
    data-testid="load-more"
  >
    <div />
  </button>
</DocumentFragment>
`;
